srv_hijacker/__init__.py

```diff
@@ -1 +1 @@
-from .srv_hijacker import hijack
+from .srv_hijacker import hijack, register, PatchError
```

Comment:
This significantly expands the public API - I don't think it's worth doing. Couldn't you just add the patch as a regular `PATCHABLE_LIB`, @sherubthakur?

Reply:
> This significantly expands the public API - I don't think it's worth doing

Elaborate.

> Couldn't you just add the patch as a regular PATCHABLE_LIB, @sherubthakur?

Totally can, but I would really prefer that we have this kind of an interface.
Here is my reasoning
- *Open for extension*: Now anyone can expand on this using a clean interface. In other words, it makes the library open for extension.
- Now if at all I want to add any lib I am kinda stuck on the library author to add/approve stuff. Basically removes the library author from being a bottleneck (There are ways to circumvent this like forking/patching this, etc but that seems to defeat the purpose of using this lib then)
- Removes need to monkey patch this lib if it is for a private repository. i.e. if I have a project which due to business needs I cannot share publicly, then the options I have are to either patch this library or fork and create a separate project.

^ Given all this I think it is worthwhile to do this.

Reply:
> This significantly expands the public API - I don't think it's worth doing

We're exposing the interface used internally for patchable libs, which includes the regex and a `resolver.Resolver()` object, which is from a third party library. This means that (a) we're exposing a third-party library to users of this library (b) we can't ever change the third-party library we use under the hood, since that'd now be a breaking change.

> Open for extension: Now anyone can expand on this using a clean interface. In other words, it makes the library open for extension.

Every line of code is a potential "extension point" in a library 🙂 The ones we'd choose to offer depends on what the primary use-case for the library is. The aim of this library is to do one thing - it changes the DNS resolution behaviour in your app to use SRV records. The core of this library is the "patches" themselves, not the framework used to call those patch functions. There's really no need to hook into this library if you're looking to add a custom patch function - just call the function yourself.

> Removes need to monkey patch this lib if it is for a private repository. i.e. if I have a project which due to business needs I cannot share publicly, then the options I have are to either patch this library or fork and create a separate project.

Sounds like what you're trying to achieve is adding another patch function - I'd suggest just calling the function yourself without hooking it into this library, I don't see much of an added benefit in calling it through this library. You still need to do the regex matching yourself and call the DNS resolver and parse results yourself. All you're basically getting from this is a function call.